core/shared/src/main/scala/zio/stm/ZSTM.scala

```diff
@@ -1040,10 +1041,20 @@ object ZSTM {
    * Applies the function `f` to each element of the `Chunk[A]` and
    * returns a transactional effect that produces a new `Chunk[B]`.
    */
-  def foreach[R, E, A, B](in: Chunk[A])(f: A => ZSTM[R, E, B]): ZSTM[R, E, Chunk[B]] =
-    in.foldLeft[ZSTM[R, E, Chunk[B]]](ZSTM.succeedNow(Chunk.empty))((acc, a) =>
-      f(a).zipWith(acc)((b, acc) => acc ++ Chunk.single(b))
-    )
+  def foreach[R, E, A, B](in: Chunk[A])(f: A => ZSTM[R, E, B]): ZSTM[R, E, Chunk[B]] = {
+    val length = in.length
+    var idx    = 0
+
+    var tx: ZSTM[R, E, ArrayBuffer[B]] = ZSTM.succeedNow(ArrayBuffer.empty[B])
+
+    while (idx < length) {
```

Comment:
I think this too could benefit from the "hook" that I mention. We are doing a lot of work that we don't need to be doing. If we can prune it down, we can dramatically increase performance of all these core operators.